src/lib/evaluate.ts

```diff
@@ -178,8 +228,129 @@ export async function evaluatePromptOnTestSuite(params: {
+    model,
+    messages: [
+      { role: 'system', content: compressingPrompt },
+      { role: 'user', content: task },
+    ],
+  });
+
+  const compressedTask = response.choices[0]?.message?.content;
+  if (!compressedTask) {
+    throw new Error('Failed to get compressed task from model.');
+  }
+  if (!response.usage) {
+    throw new Error('No usage data in compression response');
+  }
+
+  return { compressedTask, usage: response.usage };
+}
+
+/**
+ * Evaluates a compressing prompt by testing it against a test suite
+ * and calculating the compression ratio and impact on performance.
+ * @param params Parameters for evaluating the compression
+ * @returns A detailed result of the compression evaluation
+ */
+export async function evaluateCompression(params: {
+  testCases: TestCase[];
+  compressingPrompt: string;
+  compressionModel: string;
+  evaluationModel: string;
+  attemptsPerTest?: number;
+}): Promise<CompressionEvalResult> {
+  const {
+    testCases,
+    compressingPrompt,
+    compressionModel,
+    evaluationModel,
+    attemptsPerTest = 1,
+  } = params;
+
+  // 1. Baseline evaluation
+  const uncompressedResults = await evaluatePromptOnTestSuite({
+    testCases,
+    model: evaluationModel,
+    attemptsPerTest,
+  });
+
+  // 2. Compress all tasks in parallel
+  const compressionPromises = testCases.map(testCase =>
+    getCompressedTask({
+      task: testCase.task,
+      compressingPrompt,
+      model: compressionModel,
+    })
+  );
+
+  const compressionActionResults = await Promise.all(compressionPromises);
+
+  const compressedTestCases: TestCase[] = [];
+  const compressionUsage: OpenAI.CompletionUsage = {
+    prompt_tokens: 0,
+    completion_tokens: 0,
+    total_tokens: 0,
+  };
+
+  for (let i = 0; i < testCases.length; i++) {
+    const { compressedTask, usage } = compressionActionResults[i];
+    compressedTestCases.push({
+      ...testCases[i],
+      task: compressedTask,
+    });
+    compressionUsage.prompt_tokens += usage.prompt_tokens;
+    compressionUsage.completion_tokens += usage.completion_tokens;
+    compressionUsage.total_tokens += usage.total_tokens;
+  }
+
+  // 3. Evaluate compressed tasks
+  const compressedResults = await evaluatePromptOnTestSuite({
+    testCases: compressedTestCases,
+    model: evaluationModel,
+    attemptsPerTest,
+  });
+
+  // 4. Calculate metrics
+  const uncompressedPromptTokens = uncompressedResults.totalUsage.prompt_tokens;
+  const compressedPromptTokens = compressedResults.totalUsage.prompt_tokens;
+  const compressionRatio =
+    uncompressedPromptTokens > 0
+      ? (uncompressedPromptTokens - compressedPromptTokens) /
+        uncompressedPromptTokens
+      : 0;
+
+  const uncompressedPassRate =
+    uncompressedResults.passedTests / uncompressedResults.totalTests;
+  const compressedPassRate =
+    compressedResults.passedTests / compressedResults.totalTests;
+  const passRateDifference = compressedPassRate - uncompressedPassRate;
```

Comment:
We should only compute compression ratio for tests with 100% pass rate. These variables are not needed